node.go

```diff
@@ -116,16 +126,58 @@ func (n *Node) Attribute(attrID ua.AttributeID) (*ua.Variant, error) {
 	return value, nil
 }
 
-// References retrns all references for the node.
+func (n *Node) Attributes(attrID ...ua.AttributeID) ([]*ua.DataValue, error) {
+	req := &ua.ReadRequest{}
+	for _, id := range attrID {
+		rv := &ua.ReadValueID{NodeID: n.ID, AttributeID: id}
+		req.NodesToRead = append(req.NodesToRead, rv)
+	}
+	res, err := n.c.Read(req)
+	if err != nil {
+		return nil, err
+	}
+	return res.Results, nil
+}
+
+func (n *Node) Children(refs uint32, mask ua.NodeClass) ([]*Node, error) {
+	if refs == 0 {
+		refs = id.HierarchicalReferences
+	}
+	return n.ReferencedNodes(refs, ua.BrowseDirectionForward, mask, true)
+}
+
+func (n *Node) ReferencedNodes(refs uint32, dir ua.BrowseDirection, mask ua.NodeClass, includeSubtypes bool) ([]*Node, error) {
+	if refs == 0 {
+		refs = id.References
+	}
+	var nodes []*Node
+	res, err := n.References(refs, dir, mask, includeSubtypes)
+	if err != nil {
+		return nil, err
+	}
+	for _, r := range res {
+		nodes = append(nodes, n.c.Node(r.NodeID.NodeID))
```

Comment:
I wrote a browse example along the same lines to explore a little and have identified a problem with this using this pattern: `client.Node(reference.NodeID.NodeID)`.
The `ExpandedNodeId` in `reference.NodeID` may have been actually expanded and contains a namespace uri and 0 as namespace index (as per the spec). Now, assigning `reference.NodeID.NodeID` to a `Node`'s id copies the mask which contains (in my case) NamespaceUri flag and ServerIndex flag. Encoding this NodeID in the BrowseRequest writes 0xC0 as DataEncoding byte into the stream. And the Server responds with StatusBadDecodingError, because the stream doesn't contain the promised namespace URI and ServerIndex and actually the ExpandedNodeId is not allowed in BrowseRequest anyway.

You can test this by browsing the Prosys OPC UA Simulation Server in default configuration. Funny thing is, the namespace that gets expanded is the OPC UA core namespace and AFAICT it's only this namespace that gets expanded. Nevertheless, I see this as a problem with the API as the Reference.NodeID.NodeID allows one to get hold of a NodeID, which is has a mask a NodeID is not supposed to have.

A possible solution would be to make `ExpandedNodeID.nodeID` private and add a function `ExpandedNodeID.NodeID()` to safely convert to NodeID or return an error. One problem I see is one would need a lookup to resolve a namespace URI to an index.

Maybe this needs to be extracted into an issue of its own.

Reply:
ran into this exact issue yesterday

Reply:
a super quick'n'dirty work around I used:

```diff
diff --git a/node.go b/node.go
index 5adf925..00fc9fd 100644
--- a/node.go
+++ b/node.go
@@ -156,6 +156,9 @@ func (n *Node) ReferencedNodes(refs uint32, dir ua.BrowseDirection, mask ua.Node
                return nil, err
        }
        for _, r := range res {
+               r.NodeID.NodeID.ClearIndexFlag()
+               r.NodeID.NodeID.ClearURIFlag()
+
                nodes = append(nodes, n.c.Node(r.NodeID.NodeID))
        }
        return nodes, nil
diff --git a/ua/node_id.go b/ua/node_id.go
index 8533e8c..d178f47 100644
--- a/ua/node_id.go
+++ b/ua/node_id.go
@@ -179,6 +179,10 @@ func (n *NodeID) SetURIFlag() {
        n.mask |= 0x80
 }
 
+func (n *NodeID) ClearURIFlag() {
+       n.mask &^= 0x80
+}
+
 // IndexFlag returns whether the Index flag is set in EncodingMask.
 func (n *NodeID) IndexFlag() bool {
        return n.mask&0x40 == 0x40
@@ -189,6 +193,10 @@ func (n *NodeID) SetIndexFlag() {
        n.mask |= 0x40
 }
 
+func (n *NodeID) ClearIndexFlag() {
+       n.mask &^= 0x40
+}
+
 // Namespace returns the namespace id. For two byte node ids
 // this will always be zero.
 func (n *NodeID) Namespace() uint16 {
```